### PR TITLE
Implement TSTCC

### LINKED ISSUE #284: Implement TSTCC
### 🚀 Feature Request

Implement the second model for generating embeddings of ts - `TSTCCModel`
Reference [implementation](https://github.com/emadeldeen24/TS-TCC)
### Proposal

Put reference model to `libs/ts_tcc` with some changes:
- move training parameters (n_epochs, lr, temperature, lambda1, lambda2) from `init` to `fit`
- move encode parameters (encode_full_series) from `init` to `encode`

Implement:
- Class `TSTCCModel(BaseEmbeddingModel)` with methods:
  - `fit` with the same input parameters as in reference model.
  - `encode_segment` calling `encode` method of reference model with `encode_full_series=True`
  -  `encode_window` calling `encode` method of reference model with `encode_full_series=False`
  - `freeze` to enable or disable skipping training in ``fit``
  - `save` and `load`
  - property `is_freezed`

### Test cases

What should be checked:

- Fitting model
- Encoding with `encode_segment`
- Encoding with `encode_window`
- Saving pretrained model
- Loading pretrained model
  - Exception if there is no saved model
- Getting the same result when transforming dataset before saving and after loading model

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by egoriyaa**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Closing issues
closes #284

### CODE REVIEW

File: docs/source/api_reference/transforms.rst [hunk 1]
```diff
@@ -65,7 +65,8 @@ Embedding transforms and their utilities:
 
    EmbeddingSegmentTransform
    EmbeddingWindowTransform
-   embeddings.TS2VecEmbeddingModel
```
Comment (d-a-bunin) [inline docs/source/api_reference/transforms.rst:68]: Why have we made this change?

File: etna/libs/tstcc/augmentations.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+# Note: Copied from ts-tcc repository (https://github.com/emadeldeen24/TS-TCC/tree/main)
+# Fix numpy warning in `permutation` function
```
Comment (d-a-bunin) [inline etna/libs/tstcc/augmentations.py:25]: Which line? Can you explain in this thread why was it made?

File: etna/libs/tstcc/tstcc.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                total_loss.append(loss.item())
+                loss.backward()
+                model_optimizer.step()
+
+            train_loss = torch.tensor(total_loss).mean()
+
+            print(f'\nEpoch : {epoch}\n'
+                  f'Train Loss     : {train_loss:.4f}\n')
+
+        print("\n################## Training is Done! #########################")
```
Comment (d-a-bunin) [inline etna/libs/tstcc/tstcc.py:226]: Are we ok with this printing?

File: pyproject.toml [hunk 4]
```diff
@@ -245,7 +246,10 @@ filterwarnings = [
     "ignore: Trying to infer the `batch_size` from an ambiguous collection", # pytorch_forecasting
     "ignore: Implicitly cleaning up <TemporaryDirectory",
     "ignore: Call to deprecated class DeepARModel.",
-    "ignore: dropout option adds dropout after all but last recurrent layer"
+    "ignore: dropout option adds dropout after all but last recurrent layer",
+    "ignore: Using padding='same' with even kernel lengths and odd dilation may require a zero-padded copy of the input be created",
+    "ignore: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument",
```
Comment (d-a-bunin) [inline pyproject.toml:251]: Can't we get rid of this warning?

File: pyproject.toml [hunk 5]
```diff
@@ -245,7 +246,10 @@ filterwarnings = [
...
+    "ignore: Initializing zero-element tensors is a no-op"
```
Comment (d-a-bunin) [inline pyproject.toml:252]: Can't we get rid of this warning?

File: tests/test_transforms/test_inference/test_inverse_transform.py [hunk 6]
```diff
@@ -133,10 +136,26 @@ def _test_inverse_transform_train_subset_segments(self, ts, transform, segments)
                 ),
                 "regular_ts",
             ),
+            (
+                EmbeddingSegmentTransform(
```
Comment (d-a-bunin) [inline tests/test_transforms/test_inference/test_inverse_transform.py:140]: Why for `EmbeddingSegmentTransform` we have only test for `TSTCCEmbeddingModel`?

File: docs/source/api_reference/transforms.rst (hunk 1 above)
Comment (egoriyaa) [reply docs/source/api_reference/transforms.rst:68]: I think we should emphasize that it is model + we import them from `embeddings.models` everywhere in code

File: etna/libs/tstcc/augmentations.py (hunk 2 above)
Comment (egoriyaa) [reply etna/libs/tstcc/augmentations.py:25]: There this a comment in this function above needed line

File: etna/libs/tstcc/tstcc.py (hunk 3 above)
Comment (egoriyaa) [reply etna/libs/tstcc/tstcc.py:226]: image

File: tests/test_transforms/test_inference/test_inverse_transform.py (hunk 6 above)
Comment (egoriyaa) [reply tests/test_transforms/test_inference/test_inverse_transform.py:140]: we have both

File: etna/libs/tstcc/tstcc.py (hunk 3 above)
Comment (d-a-bunin) [reply etna/libs/tstcc/tstcc.py:226]: I'm not sure there should be any prints. It just clutters the output. Or we should do it using logger.

It also creates inconsistency between our transforms.

File: pyproject.toml (hunk 5 above)
Comment (egoriyaa) [reply pyproject.toml:251]: It appear in `nn.Linear(output_dims // 2, output_dims // 4)`, when set `output_dims` param < 1
I don't think we can fix it

File: etna/libs/tstcc/tstcc.py (hunk 3 above)
Comment (egoriyaa) [reply etna/libs/tstcc/tstcc.py]: When I do training in notebook, cell output is repeated like this after each following `fit` calling (first calling doesn't have such problem)
Screenshot 2024-04-08 at 12 59 09

File: pyproject.toml (hunk 5 above)
Comment (d-a-bunin) [reply pyproject.toml:251]: Which file and line is it? I can't find.
How can we set `output_dims` < 1? Why is this possible?

**Review by d-a-bunin — COMMENTED** [review]
What about changelog?

File: pyproject.toml (hunk 5 above)
Comment (egoriyaa) [reply pyproject.toml:251]: https://github.com/etna-team/etna/blob/0f77eb624337a81db4253627aee884871ff5db0b/etna/libs/tstcc/tc.py#L58

Here `input_dims` is `output_dims` in our global model.
Sorry, I mean `output_dims` < 4

**Review by d-a-bunin — APPROVED** [review]

**Review by Ama16 — APPROVED** [review]

**Pull request merged**